lupa/_lupa.pyx

```diff
@@ -1896,8 +1896,13 @@ cdef int py_args_with_gil(lua_State* L) with gil:
         finally: return -1
 
 cdef int py_args(lua_State* L) nogil:
+    cdef PyObject* runtime
+    runtime = <PyObject*>lua.lua_touserdata(L, lua.lua_upvalueindex(1))
+    if not runtime:
+        return lua.luaL_error(L, "missing runtime")
     lua.luaL_checktype(L, 1, lua.LUA_TTABLE)
-    result = py_args_with_gil(L)
+    with gil:
+        result = py_args_with_gil(<LuaRuntime>runtime, L)
```

Comment:
Just pass the `PyObject*` and cast it to `<LuaRuntime>` inside of the `with gil` function. That avoids the extra (and in the face of longjmp, risky) `with gil` block here. You can also use `<LuaRuntime?>` as a type cast to get an additional Python type check.

Reply:
Thanks!